src/main/java/org/opensearch/security/authtoken/jwt/JwtVendor.java

```diff
@@ -0,0 +1,171 @@
+        }
+        if (timeProvider.isPresent()) {
+            this.timeProvider = timeProvider.get();
+        } else {
+            this.timeProvider = () -> System.currentTimeMillis() / 1000;
+        }
+    }
+
+    /*
+     * The default configuration of this web key should be:
+     *   KeyType: OCTET
+     *   PublicKeyUse: SIGN
+     *   Encryption Algorithm: HS512
+     * */
+    static JsonWebKey createJwkFromSettings(Settings settings) throws Exception {
+        String signingKey = settings.get("signing_key");
+
+        if (!Strings.isNullOrEmpty(signingKey)) {
+
+            JsonWebKey jwk = new JsonWebKey();
+
+            jwk.setKeyType(KeyType.OCTET);
+            jwk.setAlgorithm("HS512");
+            jwk.setPublicKeyUse(PublicKeyUse.SIGN);
+            jwk.setProperty("k", signingKey);
+
+            return jwk;
+        } else {
+            Settings jwkSettings = settings.getAsSettings("jwt").getAsSettings("key");
+
+            if (jwkSettings.isEmpty()) {
+                throw new Exception(
+                    "Settings for signing key is missing. Please specify at least the option signing_key with a shared secret."
+                );
+            }
+
+            JsonWebKey jwk = new JsonWebKey();
+
+            for (String key : jwkSettings.keySet()) {
+                jwk.setProperty(key, jwkSettings.get(key));
+            }
+
+            return jwk;
+        }
+    }
+
+    public String createJwt(
+        String issuer,
+        String subject,
+        String audience,
+        Integer expirySeconds,
+        Set<String> roles,
+        Set<String> backendRoles,
+        boolean roleSecurityMode
+    ) throws Exception {
+        final long nowAsMillis = timeProvider.getAsLong();
+        final Instant nowAsInstant = Instant.ofEpochMilli(timeProvider.getAsLong());
+
+        jwtProducer.setSignatureProvider(JwsUtils.getSignatureProvider(signingKey));
+        JwtClaims jwtClaims = new JwtClaims();
+        JwtToken jwt = new JwtToken(jwtClaims);
+
+        jwtClaims.setIssuer(issuer);
+
+        jwtClaims.setIssuedAt(nowAsMillis);
+
+        jwtClaims.setSubject(subject);
+
+        jwtClaims.setAudience(audience);
+
+        jwtClaims.setNotBefore(nowAsMillis);
+
+        if (expirySeconds > maxExpirySeconds) {
+            throw new Exception("The provided expiration time exceeds the maximum allowed duration of " + maxExpirySeconds + " seconds");
+        }
+
+        expirySeconds = (expirySeconds == null) ? defaultExpirySeconds : Math.min(expirySeconds, maxExpirySeconds);
+        if (expirySeconds <= 0) {
+            throw new Exception("The expiration time should be a positive integer");
+        }
+        long expiryTime = timeProvider.getAsLong() + expirySeconds;
+        jwtClaims.setExpiryTime(expiryTime);
+
+        if (roles != null) {
+            String listOfRoles = String.join(",", roles);
+            jwtClaims.setProperty("er", encryptionDecryptionUtil.encrypt(listOfRoles));
```

Comment:
Nit: I don't think `er` and `br` are super clear. I know `br` is backend roles but still don't know what `er` means. Can we use full term or more complete abbreviation? I don't think the small size difference should be an issue

Reply:
yes the er is encrypted roles. I think this is the original design we came up with in our main branch, so lets keep using it.